pkg/snapstore/snapstore_test.go

```diff
@@ -278,10 +280,254 @@ var _ = Describe("Save, List, Fetch, Delete from mock snapstore", func() {
+								for i := range credentialFileNames {
+									// file at fileSkipIndex is skipped from being created
+									if i == fileSkippedIndex {
+										continue
+									}
+									credentialFullPaths = append(credentialFullPaths, filepath.Join(credentialDirectory, credentialFileNames[i]))
+								}
+								lastCreationTimeForFile, err := createCredentialFiles(credentialFullPaths)
+								Expect(err).ShouldNot(HaveOccurred())
+								Expect(lastCreationTimeForFile.IsZero()).ShouldNot(BeTrue())
+							})
+							It("Should return error", func() {
+								newSecretModifiedTime, err := GetSnapstoreSecretModifiedTime(providerToSnapstoreProviderMap[provider])
+								Expect(err).Should(HaveOccurred())
+								Expect(newSecretModifiedTime.IsZero()).Should(BeTrue())
+							})
+						})
+					}
+					Describe("Credential files exist", func() {
+						var lastCreationTimeForFile time.Time
+						var credentialFileFullPaths []string
+						BeforeEach(func() {
+							var err error
+							for i := range credentialFileNames {
+								credentialFileFullPaths = append(credentialFileFullPaths, filepath.Join(credentialDirectory, credentialFileNames[i]))
+							}
+							lastCreationTimeForFile, err = createCredentialFiles(credentialFileFullPaths)
+							Expect(err).ShouldNot(HaveOccurred())
+							Expect(lastCreationTimeForFile.IsZero()).ShouldNot(BeTrue())
+						})
+						Context("Files not modified", func() {
+							It("Should return creation timestamp", func() {
+								newSecretModifiedTime, err := GetSnapstoreSecretModifiedTime(providerToSnapstoreProviderMap[provider])
+								Expect(err).ShouldNot(HaveOccurred())
+								Expect(newSecretModifiedTime.Equal(lastCreationTimeForFile)).Should(BeTrue())
+							})
+						})
+						// test with each particular credential file modified
+						for fileModifiedIndex := 0; fileModifiedIndex < len(credentialFileNames); fileModifiedIndex++ {
+							fileModifiedIndex := fileModifiedIndex
+							Context("Files have been modified", func() {
+								BeforeEach(func() {
+									fmt.Println("Testing dynamic credential rotation for the provider:", providerToSnapstoreProviderMap[provider])
+									err := modifyCredentialFile(fileModifiedIndex, credentialFileFullPaths)
+									Expect(err).ShouldNot(HaveOccurred())
+								})
+								It("Should return modification timestamp", func() {
+									newSecretModifiedTime, err := GetSnapstoreSecretModifiedTime(providerToSnapstoreProviderMap[provider])
+									Expect(err).ShouldNot(HaveOccurred())
+									Expect(newSecretModifiedTime.After(lastCreationTimeForFile)).Should(BeTrue())
+								})
+							})
+						}
+					})
+					AfterEach(func() {
+						err := os.RemoveAll(credentialDirectory)
+						Expect(err).ShouldNot(HaveOccurred())
+					})
+				})
+				AfterEach(func() {
+					os.Unsetenv(provider)
+				})
+			})
+		})
+	}
+	// Credentials in JSON
+	for provider, snapstoreProvider := range providerToSnapstoreProviderJSONMap {
```

Comment:
**IMHO:**  I recommend introducing a new struct, `CredentialTestConfig`, to streamline our test cases. By defining field `CredentialType` within this struct as either "file" or "directory," we can more efficiently organize our `BeforeEach` and `AfterEach`. This approach will not only reduce redundancy but also significantly enhance the clarity and maintainability of our tests.
@renormalize WDYT ?

```go
type CredentialTestConfig struct {
	Provider        string
	EnvVariable     string
	CredentialType  string // "file" or "directory"
	CredentialFiles []string
}

// Define test entries
var credentialTestConfigs = []CredentialTestConfig{
	{
		Provider:        "AWS",
		EnvVariable:     "AWS_APPLICATION_CREDENTIALS",
		CredentialType:  "directory",
		CredentialFiles: []string{"accessKeyID", "region", "secretAccessKey"},
	},

	{
		Provider:        "AWS",
		EnvVariable:     "AWS_APPLICATION_CREDENTIALS_JSON",
		CredentialType:  "file",
		CredentialFiles: []string{"credentials.json"},
	},
}

var _ = Describe("Credential Rotation Tests", func() {

	for _, config := range credentialTestConfigs {
		Context(fmt.Sprintf("Testing credential rotation for %s with %s", config.Provider, config.EnvVariable), func() {
			Context("environment variable not set", func() {
				It("should return error", func() {

				})
			})

			Context("environment variable is set", func() {
				Context("credentials do not exist", func() {
					if config.CredentialType == "directory" {
						It("should return an error when all files are missing", func() {

						})

						for _, cfile := range config.CredentialFiles {
							It(fmt.Sprintf("should return an error when the file '%s' is missing", cfile), func() {

							})
						}
					} else {
						It("should return an error", func() {

						})
					}
				})

				Context("credentials exist", func() {
					It("should detect newly created fresh credentials", func() {
						// detecting new credentials
					})

					if config.CredentialType == "directory" {
						for _, cfile := range config.CredentialFiles {
							It(fmt.Sprintf("should detect a modified credential file '%s'", cfile), func() {

							})
						}
					} else {
						It("should detect modified credentials", func() {

						})
					}
				})
			})
		})
	}
})
```

Reply:
Addressed in [b74dff8](https://github.com/gardener/etcd-backup-restore/pull/670/commits/b74dff83631cc4d489ffc35a0d505d33d4ebc3ac)